src/drivers/magnetometer/bosch/bmm350/BMM350_registers.hpp

```diff
@@ -0,0 +1,252 @@
+ * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ ****************************************************************************/
+
+/**
+ * @file BMM350_registers.hpp
+ *
+ * Bosch BMM350 registers.
+ *
+ */
+
+#pragma once
+
+#include <cstdint>
+
+namespace Bosch_BMM350
+{
+static constexpr uint32_t I2C_SPEED          = 400000;
+static constexpr uint8_t I2C_ADDRESS_DEFAULT = 0x14;
+
+static constexpr uint8_t DEFAULT_ID          = 0x33;
+
+static constexpr int16_t OVERFLOW_XYAXES     = -4096;
+static constexpr int16_t OVERFLOW_ZAXIS      = -16384;
+
+enum class Register : uint8_t {
+	CHIP_ID           = 0x00,
+	ERR_REG           = 0x02,
+	PAD_CTRL          = 0x03,
+	PMU_CMD_AGGR_SET  = 0x04,
+	PMU_CMD_AXIS_EN   = 0x05,
+	PMU_CMD           = 0x06,
+	PMU_CMD_STATUS_0  = 0x07,
+	PMU_CMD_STATUS_1  = 0x08,
+	I3C_ERR           = 0x09,
+	I2C_WDT_SET       = 0x0A,
+	INT_CTRL          = 0x2E,
+	INT_CTRL_IBI      = 0x2F,
+	INT_STATUS        = 0x30,
+	MAG_X_XLSB        = 0x31,
+	MAG_X_LSB         = 0x32,
+	MAG_X_MSB         = 0x33,
+	MAG_Y_XLSB        = 0x34,
+	MAG_Y_LSB         = 0x35,
+	MAG_Y_MSB         = 0x36,
+	MAG_Z_XLSB        = 0x37,
+	MAG_Z_LSB         = 0x38,
+	MAG_Z_MSB         = 0x39,
+	TEMP_XLSB         = 0x3A,
+	TEMP_LSB          = 0x3B,
+	TEMP_MSB          = 0x3C,
+	SENSORTIME_XLSB   = 0x3D,
+	SENSORTIME_LSB    = 0x3E,
+	SENSORTIME_MSB    = 0x3F,
+	OTP_CMD_REG       = 0x50,
+	OTP_DATA_MSB_REG  = 0x52,
+	OTP_DATA_LSB_REG  = 0x53,
+	OTP_STATUS_REG    = 0x55,
+	TMR_SELFTEST_USER = 0x60,
+	CTRL_USER         = 0x61,
+	CMD               = 0x7E
+};
+
+// Reports Sensor Error Flag. Will be cleared on read. If the user writes a 1 into any status bit, this will also clear that bit.
+enum ERR_REG_BIT : uint8_t {
+	pmu_cmd_error = (1 << 0),           // A new PMU_CMD is issued when previous command has not been finished
+};
+
+// Configure pad behavior
+enum PAD_CTRL_DRV : uint8_t {
+	DRV_0 = 0,                          // Drive: weakest
+	DRV_2,
+	DRV_3,
+	DRV_4,
+	DRV_5,
+	DRV_6,
+	DRV_7,                              // Drive: strongest
+};
+
+// Configuration of the ODR and AVG
+enum PMU_CMD_AGGR_ODR : uint8_t {
+	ODR_400HZ    = 0x02,                // 400 Hz odr
+	ODR_200HZ    = 0x03,                // 200 Hz odr
+	ODR_100HZ    = 0x04,                // 100 Hz odr
+	ODR_50HZ     = 0x05,                // 50 Hz odr
+	ODR_25HZ     = 0x06,                // 25 Hz odr
+	ODR_12_5HZ   = 0x07,                // 12.5 Hz odr
+	ODR_6_25HZ   = 0x08,                // 6.25 Hz odr
+	ODR_3_125HZ  = 0x09,                // 3.125 Hz odr
+	ODR_1_5625HZ = 0x0A,                // 1.5625 Hz odr
+};
+
+enum PMU_CMD_AGGR_AVG : uint8_t {
+	NO_AVG = 0,                         // No average
+	AVG_2 = 1,                          // Average between 2 samples
+	AVG_4 = 2,                          // Average between 4 samples
+	AVG_8 = 3,                          // Average between 8 samples
+};
+
+union PMU_CMD_AGGR {
```

Comment:
Compiler dependent behavior, are these transferred over i2c ? If so, then I would suggest not using bit fields.

Reply:
I can add __attribute__((packed))

Reply:
It doesn't help here as there is no mention in the C standard about the storage of bitfields, their order can be chosen by the compiler at will.

Reply:
Right, cool. Removed bitfields now.